File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <vector>

using namespace std;

int M, N, K;
int ground[50][50] = {0};
//int visited[50][50] = {0};
int dy[4] = {1,-1,0,0};
int dx[4] = {0,0,1,-1};

void dfs(int y, int x){
    ground[y][x] = 0;
    
    for(int i=0; i<4; i++){
        int ddy = dy[i] + y;
        int ddx = dx[i] + x;

        if(ddy<0 || ddx<0 || ddy >= N || ddx >= M)
            continue;
        if(ground[ddy][ddx] == 1){
            //visited[y][x] = 1;
            dfs(ddy,ddx);
        }
    }

}

int main(){
    int T;
    cin >> T;

    for(int k=0;k<T;k++){
        cin >> M >> N >> K;

        int count = 0;
        ground[50][50] = {0};

        for(int i=0; i<K; i++){
            int X, Y;

            cin >> X >> Y;
            ground[Y][X] = 1;
        }

        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                if(ground[i][j] == 1){
                    //visited[i][j] = 1;
                    count ++;
                    dfs(i,j);
                }
            }
        }

        cout << count << "\n";
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {

    static int[] dy = {-1, 1, 0, 0};
    static int[] dx = {0, 0, 1, -1};

    static int t, n, m, k;
    static int[][] map;
    static boolean[][] visited;

    static boolean inRange(int y, int x){
        return (y>=0 && x>=0 && y<n && x<m);
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        t = Integer.parseInt(br.readLine());
        
        for(int tc = 0; tc<t; tc++){
            StringTokenizer st = new StringTokenizer(br.readLine());
        
            n = Integer.parseInt(st.nextToken());
            m = Integer.parseInt(st.nextToken());
            k = Integer.parseInt(st.nextToken());

            map = new int[n][m];

            for(int i=0; i<k; i++){
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                map[x][y] = 1;
            }

            visited = new boolean[n][m];

            int cnt = 0;

            for(int i=0; i<n; i++){
                for(int j=0; j<m; j++){
                    if(map[i][j] == 1 && !visited[i][j]) {
                        bfs(i,j);
                        cnt++;
                    }
                }
            }

            System.out.println(cnt);
        }
    }

    static void bfs(int y, int x){
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{y, x});
        visited[y][x] = true;

        while(!q.isEmpty()){
            int[] cur = q.poll();
            int cy = cur[0]; int cx = cur[1];

            for(int i=0; i<4; i++){
                int ddy = cy + dy[i];
                int ddx = cx + dx[i];

                if(!inRange(ddy, ddx)) continue;
                if(map[ddy][ddx] == 0) continue;
                if(visited[ddy][ddx]) continue;

                q.add(new int[]{ddy, ddx});
                visited[ddy][ddx] = true;
            }
        }
    }
}